internal/mesh/internal/types/intermediate/types.go

```diff
@@ -0,0 +1,63 @@
+package intermediate
+
+import (
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// CombinedDestinationRef contains all references we need for a specific
+// destination on the mesh.
+type CombinedDestinationRef struct {
+	// ServiceRef is the reference to the destination service.
+	ServiceRef *pbresource.Reference
+
+	// Port is the port name for this destination.
+	Port string
+
+	// SourceProxies are the reference keys of source proxy state template resources.
+	SourceProxies map[resource.ReferenceKey]struct{}
+
+	// ExplicitDestinationsID is the id of the pbmesh.Upstreams resource. For implicit destinations,
+	// this should be nil.
+	ExplicitDestinationsID *pbresource.ID
+}
+
+type ServiceEndpoints struct {
```

Comment:
I had switched off of this in https://github.com/hashicorp/consul/pull/18448 in favor of something like this:

```
type (
	DecodedDestinations     = resource.DecodedResource[pbmesh.Upstreams, *pbmesh.Upstreams]
	DecodedServiceEndpoints = resource.DecodedResource[pbcatalog.ServiceEndpoints, *pbcatalog.ServiceEndpoints]
)
```

Reply:
I'll address it in a follow-up PR to de-dup our efforts